File: 2015/22_WizardSimulator20XX/player.js
/* eslint-disable linebreak-style */
// ======================================================================
// Wizard Simulator 20XX
//   Advent of Code 2015 Day 22 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           p l a y e r . j s
//
// A player for the Advent of Code 2015 Day 22 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                 Player
// ======================================================================

class Player {
  // Player object for Wizard Simulator 20XX

  constructor(options) {
    // Create a Player object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.name = options.name === undefined ? null : options.name;
    this.hitPoints = options.hitPoints === undefined ? 0 : options.hitPoints;
    this.damage = options.damage === undefined ? 0 : options.damage;
    this.armor = options.armor === undefined ? 0 : options.armor;
    this.mana = options.mana === undefined ? 0 : options.mana;
    this.spent = options.spent === undefined ? 0 : options.spent;
    this.original = options.original === undefined ? null : options.original;

    // 2. Process text (if any)
    if (this.text !== null) {
      this.processText(this.text);
      this.text = null;
    }

    // 3. Save initial values
    if (this.original === null) {
      this.original = {
        hitPoints: this.hitPoints,
        damage: this.damage,
        armor: this.armor,
        mana: this.mana,
      };
    }
  }

  processText(text) {
    // 1. Start with nothing
    this.hp = 0;
    this.damage = 0;
    this.armor = 0;
    this.mana = 0;
    // 2. Loop for all line of the text
    text.forEach((line) => {
      // 3. Save section name if this is a section line?
      const [what, value] = line.split(':');
      const intValue = parseInt(value, 10);
      switch (what) {
        case 'Hit Points':
          this.hitPoints = intValue;
          break;
        case 'Damage':
          this.damage = intValue;
          break;
        case 'Armor':
          this.armor = intValue;
          break;
        case 'Mana':
          this.mana = intValue;
          break;
        default:
          // eslint-disable-next-line no-console
          console.log(`Unkown type for ${line}`);
      }
    });
  }

  reset() {
    this.hitPoints = this.original.hitPoints;
    this.damage = this.original.damage;
    this.armor = this.original.armor;
    this.mana = this.original.mana;
  }

  heal(points) {
    this.hitPoints += points;
  }

  protect(points) {
    this.armor += points;
  }

  unprotect() {
    this.armor = 0;
  }

  recharge(points) {
    this.mana += points;
  }

  cast(cost) {
    // 1. If we have tha mana for the spell ...
    if (cost < this.mana) {
      // 2. Deduct it from our pool, and record the expenditure
      this.mana -= cost;
      this.spent += cost;
    } else {
      // 3. Else, we were being very naughty and need to be chastised
      this.hitPoints = 0;
    }
  }

  defend(points) {
    let damageTaken = points - this.armor;
    if (damageTaken <= 0) {
      damageTaken = 1;
    }
    this.hitPoints -= damageTaken;
    if (this.hitPoints < 0) {
      this.hitPoints = 0;
    }
  }

  attack(other) {
    other.defend(this.damage);
  }
}

module.exports.Player = Player;
// ======================================================================
// end                       p l a y e r . j s                        end
// ======================================================================

File: 2015/22_WizardSimulator20XX/player.py
# ======================================================================
# Wizard Simulator 20XX
#   Advent of Code 2015 Day 22 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         p l a y e r . p y
# ======================================================================
"Player for the Advent of Code 2015 Day 22 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                 Player
# ======================================================================


class Player(object):   # pylint: disable=R0902, R0205
    "Object for Wizard Simulator 20XX"

    def __init__(self, name="noone", text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.name = name
        self.attributes = {
            'hitpoints': 0,
            'damage': 0,
            'armor': 0,
            'mana': 0,
            'used': 0,
        }

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 1. Loop for all the lines
        for line in text:

            # 2. Split line into attribute and number
            attribute, number = line.split(':')

            # 3. Save the number as hitpoints, damage, or armor
            self.attributes[attribute.lower().replace(' ', '')] = int(number)

    def is_dead(self):
        "Return True is the player is an ex-Player"
        return self.attributes['hitpoints'] <= 0

    def defend(self, attack):
        "Take damage based on the attack minus armor"
        self.attributes['hitpoints'] -= max(1, attack - self.attributes['armor'])
        self.attributes['hitpoints'] = max(0, self.attributes['hitpoints'])
        return self.is_dead()

    def wack(self):
        "Take a single point of damage"
        self.attributes['hitpoints'] -= 1

    def attack(self):
        "Returns the player's damage rating"
        return self.attributes['damage']

    def clone(self):
        "Make a duplicate player"

        # 1. Start with a clean player
        result = Player()

        # 2. Copy the fields
        result.text = self.text
        result.name = self.name
        result.part2 = self.part2
        for key, value in self.attributes.items():
            result.attributes[key] = value

        # 3. Return the copy
        return result

    def __getitem__(self, key):
        return self.attributes[key]

    def __setitem__(self, key, value):
        self.attributes[key] = int(value)

    def __str__(self):
        non_zero = ["%s:%d" % (_[0], self.attributes[_]) for _ in self.attributes
                    if self.attributes[_] > 0]
        return "%s(%s)" % (self.name[0], ",".join(non_zero))


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        p l a y e r . p y                       end
# ======================================================================
